feat(web): workflows allow searching by name or trigger identifier

LetItRock opened the pull request:
### What change does this PR introduce?

Workflows page: 
- added the search input component at the top of the workflows list
- allow searching by the name or trigger identifier
- clear functionality
- `query` parameter in the URL along with pagination params

### Why was this change needed?

<!-- If your PR fixes an open issue, use `Closes #999` to link your PR with the issue. #999 stands for the issue number you are fixing, Example: Closes #31 -->

### Other information (Screenshots)

https://github.com/novuhq/novu/assets/2607232/ca80309a-c4e7-4bae-a913-858d0aed1bd2

LetItRock commented on apps/web/src/components/layout/components/OrganizationSelect.tsx:
```diff
@@ -42,7 +42,7 @@ export default function OrganizationSelect() {
       const token = await changeOrganization(organizationId);
       setToken(token);
-      await queryClient.refetchQueries();
+      await queryClient.clear();
```
when switching the organization in the dropdown we don't need to refetch the queries but rather clear them all, otherwise some of the queries might be in the fetching state resulting in no loading indicators for example on the workflows page

LetItRock commented on apps/web/src/hooks/usePaginatedQuery.ts:
```diff
@@ -34,13 +35,16 @@ export const usePaginatedQuery = <TResponse extends object>({
   paginationOptions = {},
 }: UsePaginatedQueryProps<TResponse>) => {
   const { currentPageNumber, pageSize, setCurrentPageNumber, setPageSize } = usePaginationState(paginationOptions);
+  const { search, setSearch } = useSearchState(paginationOptions);
```
created the `useSearchState` hook that reads the `query` value from the URL query params and allows to updating it

LetItRock commented on apps/web/src/pages/templates/WorkflowListPage.tsx:
```diff
@@ -156,15 +157,18 @@ function WorkflowListPage() {
-    pageSize,
   } = useTemplates({ areSearchParamsEnabled: true });
-  const isLoading = areNotificationGroupLoading || loading;
+  const [searchValue, setSearchValue] = useState(search ?? '');
```
the search value state for the input to reflect quickly what user types; the request, and URL updates are debounced

LetItRock commented on apps/web/src/pages/templates/WorkflowListPage.tsx:
```diff
@@ -199,15 +205,28 @@ function WorkflowListPage() {
     navigate(parseUrl(ROUTES.WORKFLOWS_EDIT_TEMPLATEID, { templateId: row.values._id }));
   }
 
+  const debouncedSearchChange = useDebounce((value: string) => setSearch(value), 2000);
```
the time amount is debatable

LetItRock commented on apps/web/src/pages/templates/WorkflowListPage.tsx:
```diff
@@ -199,15 +205,28 @@ function WorkflowListPage() {
 
+  const debouncedSearchChange = useDebounce((value: string) => setSearch(value), 2000);
+
+  const onSearchClearClick = () => {
+    debouncedSearchChange.cancel();
+    setSearchValue('');
+    setSearch('');
+  };
```
on input clear button click - cancel the last debounced request and update value and make request right away

LetItRock commented on apps/web/src/pages/templates/WorkflowListPage.tsx:
```diff
@@ -199,15 +205,28 @@ function WorkflowListPage() {
+    setSearch('');
+  };
+
+  const onSearchChange: ChangeEventHandler<HTMLInputElement> = ({ target: { value } }) => {
+    if (value === '') {
+      onSearchClearClick();
+
+      return;
+    }
+    setSearchValue(value);
+    debouncedSearchChange(value);
+  };
```
on search change, update the state for input and debounce the request

LetItRock commented on apps/web/src/pages/templates/WorkflowListPage.tsx:
```diff
@@ -221,72 +240,50 @@ function WorkflowListPage() {
       }}
     >
       <Container fluid sx={{ padding: '0 24px 8px 24px' }}>
-        {isTemplateStoreEnabled ? (
```
removed the template store FF, as we use it with `true` option for a while now

LetItRock commented on apps/web/src/pages/templates/WorkflowListPage.tsx:
```diff
@@ -221,72 +240,50 @@ function WorkflowListPage() {
+            onTemplateClick={handleOnBlueprintClick}
+            onAllTemplatesClick={openModal}
+          />
+          <SearchInput
+            value={searchValue}
+            placeholder="Type name or identifier..."
+            onChange={onSearchChange}
+            onClearClick={onSearchClearClick}
+            disabled={isSearchInputDisabled}
+            data-test-id="workflows-search-input"
+          />
```
the search input added

LetItRock commented on libs/design-system/src/pagination/usePaginationState.ts:
```diff
@@ -35,13 +34,11 @@ export const usePaginationState = ({
-      },
-      { replace: true }
-    );
+    const newSearchParams = new URLSearchParams(document.location.search);
+    newSearchParams.set(pageNumberParamName, `${currentPageNumber}`);
+    newSearchParams.set(pageSizeParamName, `${pageSize}`);
+
+    setSearchParams(newSearchParams, { replace: true });
```
the `usePaginationState` and `useSearchState` are both separate hooks that both update the query parameters. This code ensures that they are not overridden but all combined together. Otherwise, if the `useEffect` in the `useSearchState` hook is called after, it will override all the pagination query params.

antonjoel82 reviewed (changes requested):
Looks good overall! Nice job on separating the rendering from the logic, and making the logic reusable.

Code thoughts:
- Should we add a minimum string length before we fire the query? Generally I see 3 char min

Testing feedback:
- What is the expected behavior if the user types an illegal character? Currently it removes the entire search string. I feel like it would be better to just omit the illegal chars.
- We should change the `search` icon color when the input is disabled
- Query param stays even when empty (i.e. `...?query=`) --> could we clear it for better UX?
- Should we convert all query strings to lower case (in the URL)?
- Changing between environments removes the query string, but it doesn't between orgs

Design feedback (I think we can make a decision here, and consult Nik when he returns):
- I think the input will feel / look nicer if we have an active state that adds either a border or background color.
- We could add a hover state (to make it more discoverable), and use the same style as above

As for the flickering behavior you mentioned on Slack, I unfortunately think this is because our `WorkflowListPage` is monolithic and over-complex, so everything re-renders together on Organization change.

<img width="581" alt="Screenshot 2024-03-08 at 12 03 11 PM" src="https://github.com/novuhq/novu/assets/47441786/00cd65e4-9801-4c36-8a75-6552f718b272">

I suggest we acknowledge that minor visual glitch as acceptable for now (since users aren't going to rapidly click between orgs), and aim to tackle it in a future refactor that aims to break apart our larger pages into more performant, reusable patterns.

LetItRock commented on libs/design-system/src/search-input/useSearchState.ts:
```diff
@@ -0,0 +1,43 @@
+import { useCallback, useEffect, useState } from 'react';
+import { useSearchParams } from 'react-router-dom';
+
+const URL_PARAM_NAME_SEARCH = 'query';
```
I thought about that as well, but then wanted to align it more with API params, but there is no need actually.

LetItRock commented:
Thanks for the review! 🙌 

> * Should we add a minimum string length before we fire the query? Generally I see 3 char min

There was no specific requirement on this from the product. I see that it might be even useful with 2 chars, but for now, I would say let's leave it like this.

> * What is the expected behavior if the user types an illegal character? Currently it removes the entire search string. I feel like it would be better to just omit the illegal chars.

All the characters are decoded in the URL, and we also don't have any restrictions on the API side for the workflow name, so the users can even call it `%test`. 
Can you provide an example? I've fixed one decoding issue that I introduced, but I haven't seen that the entire search was removed.

> * We should change the `search` icon color when the input is disabled

Mantine is not propagating the opacity on the icons :/ will add css manually

> * Query param stays even when empty (i.e. `...?query=`) --> could we clear it for better UX?

Yes, will update it.

> * Should we convert all query strings to lower case (in the URL)?

There is no need, as the API searches the DB case-insensitively. 

> * Changing between environments removes the query string, but it doesn't between orgs

The environment switch change redirects the user to the `/` => `/workflows` page, which basically resets all query parameters. The organization switch doesn't do that, but I think it should behave the same. Now it leads to the issue that the params are preserved and the user might not see the workflows if the new org doesn't have that many pages.

> Design feedback (I think we can make a decision here, and consult Nik when he returns):
> 
> * I think the input will feel / look nicer if we have an active state that adds either a border or background color.
> * We could add a hover state (to make it more discoverable), and use the same style as above
> 
> As for the flickering behavior you mentioned on Slack, I unfortunately think this is because our `WorkflowListPage` is monolithic and over-complex, so everything re-renders together on Organization change.

Yes, I agree it would be nice to have these micro-interactions on the search input so the user will have a better experience. Let's talk to Nikolay about that when he will be back.

About the `WorkflowListPage` yes it's gigantic and screams for the refactoring.

antonjoel82 commented on apps/web/src/api/notification-templates.ts:
```diff
@@ -1,11 +1,19 @@
-export function getNotificationsList(page = 0, limit = 10) {
-  return api.getFullResponse(`/v1/notification-templates`, { page, limit });
+export function getNotificationsList({ page = 0, limit = 10, query }: IPaginationWithQueryParams) {
+  const params = { page, limit, ...(query && { query }) };
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❔ question:** Is there another way we could write the query portion?‏

LetItRock replied:
if statement or ternary

antonjoel82 commented on libs/design-system/src/search-input/SearchInput.tsx:
```diff
@@ -0,0 +1,61 @@
+export interface ISearchInputProps {
+  placeholder?: string;
+  value?: string;
+  ref?: React.Ref<HTMLInputElement>;
```
Ah, so is it meant to represent the type of the ref-forwarded component then?

LetItRock replied:
exactly :)

antonjoel82 reviewed (approved):
Awesome, love the changes! Thank you for reusing interfaces from `shared` as well -- I think that's a great addition.

Everything looked good in my smoke testing 🎉
